1006 upload camera lecture slides as vod

mono424 opened the pull request:
### Motivation and Context
Currently its not possible to add lecture slides and camera videos as dedicated video files.

### Description
Add input fields to allow the upload of camera and slides video too. Because the upload formular becomes a little
user unfriendly, it is refactored. Also, a jump to new added lectures and is introduced together with a little marker
that marks the new added lectures in the list.

### Steps for Testing
Prerequisites:
- 1 Admin
- 1 Course

1. Log in
2. Navigate to a Course Admin Page
3. Upload a new VoD with dedicated camera and slides video
4. Check if all videos are working
5. Upload a new VoD with combined stream only
6. Check if the VoD with combined stream is working

### Screenshots

https://github.com/joschahenningsen/TUM-Live/assets/17506411/7052c833-3b1d-4a25-acbe-c602929107d2

mono424 committed:
started redesigning vod upload

mono424 committed:
type select buttons finifshed

mono424 committed:
steps active color

mono424 committed:
more design changes

mono424 committed:
back button

mono424 committed:
finished last slide

mono424 committed:
:)

mono424 committed:
fixed change date

mono424 committed:
:)

mono424 committed:
scroll to new created lectures

mono424 committed:
Add videotype to upload key; Introduce uploadVODMedia endpoint

mono424 committed:
Added VideoType to grpc token-info

mono424 committed:
Return streamId on upload

mono424 committed:
Works :)

mono424 committed:
Fix progress ui

mono424 committed:
updated dao

joschahenningsen commented on web/template/partial/course/manage/create-lecture-form.gohtml:
```diff
@@ -1,127 +1,216 @@
-                       :required="!formData.vodup"/>
-            </label>
-        </template>
+            <div x-show="currentTab === 0" class="flex py-4">
```
These should probably be templates to avoid weird behavior like the date picker being visible when clicking back

mono424 replied:
@joschahenningsen I tried templates, but the date picker looses the state on going back then, I added `.prevent` to the next and back button. I think that resolved the issue you are experiencing with the date picker :)

joschahenningsen commented on web/template/partial/course/manage/create-lecture-form.gohtml:
```diff
@@ -1,127 +1,216 @@
+                <ul class="w-full">
+                    <li class="mb-2">
+                        <button
+                            @click="() => updateType(false)"
```
IMO it would be nice if the page automatically switched to the second page without clicking continue here.

joschahenningsen commented on web/ts/edit-course.ts:
```diff
@@ -525,13 +556,76 @@ export function createLectureForm(args: { s: [] }) {
+            }, 100);
+        },
+        onUpdate() {
+            if (this.currentTab === 0) {
+                this.canContinue = true;
+                this.canGoBack = false;
+                this.onLastSlide = false;
+                return;
+            }
+
+            if (this.currentTab === 1) {
+                if (this.formData.vodup) {
+                    this.canGoBack = true;
+                    this.onLastSlide = false;
+                    this.canContinue = this.formData.title.length > 0 && this.formData.start.length > 0;
+                } else {
+                    this.canGoBack = true;
+                    this.onLastSlide = true;
+                    this.canContinue =
+                        this.formData.title.length > 0 &&
+                        this.formData.start.length > 0 &&
+                        this.formData.end.length > 0;
+                }
+                return;
```
This could use some comments

joschahenningsen commented on web/ts/edit-course.ts:
```diff
@@ -525,13 +556,76 @@ export function createLectureForm(args: { s: [] }) {
+                    this.canGoBack = true;
+                    this.onLastSlide = true;
+                    this.canContinue =
+                        this.formData.title.length > 0 &&
+                        this.formData.start.length > 0 &&
```
A title should not be required

joschahenningsen commented on web/ts/edit-course.ts:
```diff
@@ -525,13 +556,76 @@ export function createLectureForm(args: { s: [] }) {
+                if (this.formData.vodup) {
+                    this.canGoBack = true;
+                    this.onLastSlide = false;
+                    this.canContinue = this.formData.title.length > 0 && this.formData.start.length > 0;
```
A title should not be required

joschahenningsen commented on web/ts/edit-course.ts:
```diff
@@ -620,44 +727,93 @@ export function createLectureForm(args: { s: [] }) {
-                );
+
+        async uploadVod(): Promise<number> {
+            const uploadPres = this.formData.presFile[0] != null;
+            const uploadCam = this.formData.camFile[0] != null;
```
COMB should not be required, the only requirement should be `at least one file is present`

joschahenningsen commented on web/ts/edit-course.ts:
```diff
@@ -620,44 +727,93 @@ export function createLectureForm(args: { s: [] }) {
+
+            const { streamID } = JSON.parse(
+                await uploadFilePost(
+                    `/api/course/${this.courseID}/uploadVOD?start=${this.formData.start}&title=${this.formData.title}`,
+                    this.formData.combFile[0],
+                    (progress) => {
+                        uploadProgress = { ...uploadProgress, COMB: progress };
+                        window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
+                    },
+                ),
             );
-            xhr.send(vodUploadFormData);
+
+            if (uploadPres)
+                await uploadFilePost(
+                    `/api/course/${this.courseID}/uploadVODMedia?streamID=${streamID}&videoType=PRES`,
+                    this.formData.presFile[0],
+                    (progress) => {
+                        uploadProgress = { ...uploadProgress, PRES: progress };
+                        window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
+                    },
+                );
+
+            if (uploadCam)
+                await uploadFilePost(
+                    `/api/course/${this.courseID}/uploadVODMedia?streamID=${streamID}&videoType=CAM`,
+                    this.formData.camFile[0],
+                    (progress) => {
+                        uploadProgress = { ...uploadProgress, CAM: progress };
+                        window.dispatchEvent(new CustomEvent("voduploadprogress", { detail: uploadProgress }));
```
Can you refactor this to its own method to avoid duplication?

mono424 replied:
not sure how to do it in a clean way in a function, because I have the `uploadProgress` state that is caried through the calls, but I abstracted in a way that I have an array of mediaUploads now and iterate through, so the `uploadFilePost`-Part is there only once. 👍

joschahenningsen commented on api/courses.go:
```diff
@@ -363,7 +369,78 @@ func (r coursesRoutes) uploadVOD(c *gin.Context) {
+	p.ServeHTTP(c.Writer, c.Request)
+}
+
+func (r coursesRoutes) uploadVODMedia(c *gin.Context) {
```
Why do we need another handler? Can you elaborate & add a comment what this is used for?

mono424 replied:
My intention was to create an endpoint that lets a user upload VOD Media to an existing VOD without the need of recreating the VOD. This could be used to edit the video on demand media files later too.

joschahenningsen commented on api/courses.go:
```diff
@@ -1113,6 +1190,12 @@ func (r coursesRoutes) createLecture(c *gin.Context) {
 	seriesIdentifier := uuid.NewV4().String()
 	req.DateSeries = append(req.DateSeries, req.Start)
 
+	// To get new ids after insert
+	var oldStreamIds []uint
+	for _, stream := range tumLiveContext.Course.Streams {
+		oldStreamIds = append(oldStreamIds, stream.ID)
+	}
+
```
What exactly does this do?

mono424 replied:
I save the old ids of the streams of a course, to find the newly created stream ids after the insert. I use them to mark as newly created in the frontend.
There could not figure out how to get back the newly created ids from `CoursesDao.UpdateCourse`. Is there a way to easily get the newly created stream ids from `CoursesDao.UpdateCourse` ?

joschahenningsen commented on model/upload-key.go:
```diff
@@ -2,11 +2,24 @@ package model
 
 import "gorm.io/gorm"
 
+type VideoType string
+
+const (
+	VideoTypeCombined     VideoType = "COMB"
+	VideoTypePresentation           = "PRES"
+	VideoTypeCamera                 = "CAM"
+)
+
+func (v VideoType) Valid() bool {
+	return v == VideoTypeCombined || v == VideoTypePresentation || v == VideoTypeCamera
+}
+
```
Nice way of doing this! Unsure about this: did you check whether the stream.go model already has a similar enum?

mono424 replied:
Searched for such an enum, but I did not find any :(

joschahenningsen commented on web/template/partial/course/manage/create-lecture-form.gohtml:
```diff
@@ -1,127 +1,216 @@
-            </select>
-        </label>
+        <div class="form-container-body">
+            <div x-show="currentTab === -1" x-data="{ progress: { COMB: 0, PRES: null, CAM: null } }" @voduploadprogress.window="(e) => (progress = e.detail)">
```
The tabs could potentially be in their own templates if that makes things a little easier to understand

joschahenningsen reviewed (changes requested):
Amazing! Left a few thoughts :) :clinking_glasses:

mono424 committed:
first iteration of feedback impl.

mono424 committed:
uncoupled creation and upload

mono424 committed:
linter

mono424 committed:
cleanup on error

mono424 committed:
splitted by tabs

mono424 committed:
Merge branch 'dev' into 1006-upload-camera-lecture-slides-as-vod

mono424 commented:
@joschahenningsen I did some major changes due to your feedback, I hope it makes it simpler. Summary:
I renamed `uploadVOD` to `createVOD` and removed the video upload logic. All videos are not uploaded over the dedicated endpoint `uploadVODMedia`. Now the upload logic only has to be in one of the two endpoints. Also the worker does not have to return the streamID but it can be directly returned from `createVOD`.
Downside, you could theoretically create a vod without any media if you sent the request manually to the API. But thats probably not to bad.

mono424 committed:
fix js

mono424 committed:
remove local workaround

mono424 committed:
remove optional text

mono424 committed:
revert local workaround

joschahenningsen reviewed (approved):
Looks good! Thanks :clinking_glasses:
